src/many-migration/src/lib.rs

```diff
@@ -0,0 +1,478 @@
+        update_fn: FnPtr<T, E>,
+        name: &'static str,
+        description: &'static str,
+    ) -> Self {
+        Self {
+            r#type: MigrationType::Regular(RegularMigration {
+                initialize_fn,
+                update_fn,
+            }),
+            name,
+            description,
+        }
+    }
+
+    pub const fn new_initialize(
+        initialize_fn: FnPtr<T, E>,
+        name: &'static str,
+        description: &'static str,
+    ) -> Self {
+        Self {
+            r#type: MigrationType::Regular(RegularMigration {
+                initialize_fn,
+                update_fn: |_| Ok(()),
+            }),
+            name,
+            description,
+        }
+    }
+
+    pub const fn new_update(
+        update_fn: FnPtr<T, E>,
+        name: &'static str,
+        description: &'static str,
+    ) -> Self {
+        Self {
+            r#type: MigrationType::Regular(RegularMigration {
+                initialize_fn: |_| Ok(()),
+                update_fn,
+            }),
+            name,
+            description,
+        }
+    }
+}
+
+impl<'a, T, E> InnerMigration<'a, T, E> {
+    pub const fn name(&self) -> &'a str {
+        self.name
+    }
+
+    pub const fn description(&self) -> &'a str {
+        self.description
+    }
+
+    pub const fn r#type(&self) -> &'_ MigrationType<T, E> {
+        &self.r#type
+    }
+
+    /// This function gets executed when the storage block height == the migration block height
+    pub fn initialize(&self, storage: &mut T) -> Result<(), E> {
+        match &self.r#type {
+            MigrationType::Regular(migration) => (migration.initialize_fn)(storage),
+            _ => {
+                tracing::trace!(
+                    "Migration {} is not of type `Regular`, skipping",
+                    self.name()
+                );
+                Ok(())
+            }
+        }
+    }
+
+    /// This function gets executed when the storage block height >= the migration block height
+    pub fn update(&self, storage: &mut T) -> Result<(), E> {
+        match &self.r#type {
+            MigrationType::Regular(migration) => (migration.update_fn)(storage),
+            _ => {
+                tracing::trace!(
+                    "Migration {} is not of type `Regular`, skipping",
+                    self.name()
+                );
+                Ok(())
+            }
+        }
+    }
+
+    /// This function gets executed when the storage block height == the migration block height
+    pub fn hotfix<'b>(&'b self, b: &'b [u8]) -> Option<Vec<u8>> {
+        match &self.r#type {
+            MigrationType::Hotfix(migration) => (migration.hotfix_fn)(b),
+            _ => {
+                tracing::trace!(
+                    "Migration {} is not of type `Hotfix`, skipping",
+                    self.name()
+                );
+                None
+            }
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct IO<'a> {
```

Comment:
I really don't like `IO`, since it has little to do with input/output. Maybe `MigrationData` or `MigrationJson`?

Also, acronyms should be camel-cased, so `Io` in this case, or `Json` (not `JSON`).